### PR TITLE
Delete messages and conversations

### PR DESCRIPTION
**Pull request by oldmud0**
Ann accidentally converted spaces to tabs, realized her mistake, and converted it back. Sorry for that.

### CODE REVIEW

File: src/codeu/chat/client/core/Controller.java [hunk 1]
```diff
@@ -239,4 +239,53 @@ public void setConversationExplicitPermissions(Uuid convoID, Uuid invoker, Uuid
...
+	        LOG.info("DeleteMessage: Response completed.");
+	      } else if (returnCode == NetworkCode.ERR_SECURITY_VIOLATION) {
+	        throw new SecurityViolationException();
+	      } else {
+	        LOG.error("Response from server failed.");
+	      }
+	  } catch (SecurityViolationException e) {
+	    throw e;
+	  } catch (Exception ex) {
+		  System.out.println("ERROR: Exception during call on server. Check log for details.");
```
Comment (cbick) [inline src/codeu/chat/client/core/Controller.java]: indentation's a little funky here

File: src/codeu/chat/common/BasicView.java [hunk 2]
```diff
@@ -62,5 +62,7 @@
 
   String getAllConvosFromServer(Uuid signedInId);
   String getNumMessagesFromServer(Uuid signedInId);
+
```
Comment (cbick) [inline src/codeu/chat/common/BasicView.java]: revert this file, only has whitespace changes

File: src/codeu/chat/contexts/MessageContext.java [hunk 3]
```diff
@@ -19,6 +19,7 @@
 
 import codeu.chat.common.BasicView;
 import codeu.chat.common.Message;
+import codeu.chat.security.SecurityViolationException;
```
Comment (cbick) [inline src/codeu/chat/contexts/MessageContext.java]: doesn't seem like this gets used?

File: src/codeu/chat/client/core/Controller.java [hunk 4]
```diff
@@ -239,4 +239,53 @@ public void setConversationExplicitPermissions(Uuid convoID, Uuid invoker, Uuid
...
+	      if (returnCode == NetworkCode.DELETE_MESSAGE_RESPONSE) {
+	        LOG.info("DeleteMessage: Response completed.");
+	      } else if (returnCode == NetworkCode.ERR_SECURITY_VIOLATION) {
+	        throw new SecurityViolationException();
+	      } else {
+	        LOG.error("Response from server failed.");
+	      }
+	  } catch (SecurityViolationException e) {
+	    throw e;
+	  } catch (Exception ex) {
```
Comment (cbick) [inline src/codeu/chat/client/core/Controller.java]: this pattern of catching stray exceptions can cause problems. is there a particular set of exceptions you're expecting to happen here? if you're just catching bugs, it'd be better to let them propagate (and get fixed)

File: src/codeu/chat/server/contexts/UserContext.java [hunk 5]
```diff
@@ -69,4 +70,21 @@ public String removeConvoInterest(String name) {
...
 
+  @Override
+  public void deleteConversation(Uuid conversationId) throws SecurityViolationException {
+    ConversationContext foundConversation = null;
+    for (ConversationContext convo : conversations()) {
+      if (convo.conversation.id.equals(conversationId)) {
+        foundConversation = convo;
+      }
+    }
+    // HACK: I want to throw an IllegalArgumentException but without changing the method signature
```
Comment (cbick) [inline src/codeu/chat/server/contexts/UserContext.java]: IAE is a runtime exception, it should be fine to throw it regardless.

File: src/codeu/chat/server/contexts/UserContext.java [hunk 6]
```diff
@@ -69,4 +70,21 @@ public String removeConvoInterest(String name) {
...
+    if (foundConversation == null) return;
```
Comment (cbick) [inline src/codeu/chat/server/contexts/UserContext.java]: IMO it makes sense to throw security violation even if it doesn't exist (but it's fine if it doesn't)

File: test/codeu/chat/util/store/StoreTest.java [hunk 7]
```diff
@@ -147,6 +147,35 @@ public void testFirst() {
     assertTrue(store.first(4) == 40);
   }
 
+  @Test
```
Comment (cbick) [inline test/codeu/chat/util/store/StoreTest.java:150]: yay tests!

**Review by cbick — CHANGES_REQUESTED** [review]

File: src/codeu/chat/client/core/Controller.java (hunk 4 above)
Comment (oldmud0) [reply src/codeu/chat/client/core/Controller.java]: We were following the (anti-)pattern of the original code, of logging the exceptions that could possibly be thrown during I/O (broken pipe, connection reset, etc.) with a blanket `Exception`. As you say, I will change it to an `IOException`.

File: src/codeu/chat/client/core/Controller.java (hunk 1 above)
Comment (oldmud0) [reply src/codeu/chat/client/core/Controller.java]: Ann used tab indents for some reason. I pointed it out to her, but it seems she did not correct it everywhere.

File: src/codeu/chat/contexts/MessageContext.java (hunk 3 above)
Comment (oldmud0) [reply src/codeu/chat/contexts/MessageContext.java]: There are unused imports everywhere. I'll get them all removed.

**Pull request merged**